[ML] Improve file structure finder timestamp format determination

droberts195 opened the pull request:
This change contains a major refactoring of the timestamp
format determination code used by the ML find file structure
endpoint.

Previously timestamp format determination was done separately
for each piece of text supplied to the timestamp format finder.
This had the drawback that it was not possible to distinguish
dd/MM and MM/dd in the case where both numbers were 12 or less.
In order to do this sensibly it is best to look across all the
available timestamps and see if one of the numbers is greater
than 12 in any of them.  This necessitates making the timestamp
format finder an instantiable class that can accumulate evidence
over time.

Another problem with the previous approach was that it was only
possible to override the timestamp format to one of a limited
set of timestamp formats.  There was no way out if a file to be
analysed had a timestamp that was sane yet not in the supported
set.  This is now changed to allow any timestamp format that can
be parsed by a combination of these Java date/time formats:
yy, yyyy, M, MM, MMM, MMMM, d, dd, EEE, EEEE, H, HH, h, mm, ss,
a, XX, XXX, zzz
Additionally S letter groups (fractional seconds) are supported
providing they occur after ss and separated from the ss by a dot,
comma or colon.  Spacing and punctuation is also permitted with
the exception of the question mark, newline and carriage return
characters, together with literal text enclosed in single quotes.

The full list of changes/improvements in this refactor is:

- Make TimestampFormatFinder an instantiable class
- Overrides must be specified in Java date/time format - Joda
  format is no longer accepted
- Joda timestamp formats in outputs are now derived from the
  determined or overridden Java timestamp formats, not stored
  separately
- Functionality for determining the "best" timestamp format in
  a set of lines has been moved from TextLogFileStructureFinder
  to TimestampFormatFinder, taking advantage of the fact that
  TimestampFormatFinder is now an instantiable class with state
- The functionality to quickly rule out some possible Grok
  patterns when looking for timestamp formats has been changed
  from using simple regular expressions to the much faster
  approach of using the Shift-And method of sub-string search,
  but using an "alphabet" consisting of just 1 (representing any
  digit) and 0 (representing non-digits)
- Timestamp format overrides are now much more flexible
- Timestamp format overrides that do not correspond to a built-in
  Grok pattern are mapped to a %{CUSTOM_TIMESTAMP} Grok pattern
  whose definition is included within the date processor in the
  ingest pipeline
- Grok patterns that correspond to multiple Java date/time
  patterns are now handled better - the Grok pattern is accepted
  as matching broadly, and the required set of Java date/time
  patterns is built up considering all observed samples
- As a result of the more flexible acceptance of Grok patterns,
  when looking for the "best" timestamp in a set of lines
  timestamps are considered different if they are preceded by
  a different sequence of punctuation characters (to prevent
  timestamps far into some lines being considered similar to
  timestamps near the beginning of other lines)
- Out-of-the-box Grok patterns that are considered now include
  %{DATE} and %{DATESTAMP}, which have indeterminate day/month
  ordering
- The order of day/month in formats with indeterminate day/month
  order is determined by considering all observed samples (plus
  the server locale if the observed samples still do not suggest
  an ordering)

Relates #38086
Closes #35137
Closes #35132

droberts195 committed:
[ML] Improve file structure finder timestamp format determination

This change contains a major refactoring of the timestamp
format determination code used by the ML find file structure
endpoint.

Previously timestamp format determination was done separately
for each piece of text supplied to the timestamp format finder.
This had the drawback that it was not possible to distinguish
dd/MM and MM/dd in the case where both numbers were 12 or less.
In order to do this sensibly it is best to look across all the
available timestamps and see if one of the numbers is greater
than 12 in any of them.  This necessitates making the timestamp
format finder an instantiable class that can accumulate evidence
over time.

Another problem with the previous approach was that it was only
possible to override the timestamp format to one of a limited
set of timestamp formats.  There was no way out if a file to be
analysed had a timestamp that was sane yet not in the supported
set.  This is now changed to allow any timestamp format that can
be parsed by a combination of these Java date/time formats:
yy, yyyy, M, MM, MMM, MMMM, d, dd, EEE, EEEE, H, HH, h, mm, ss,
a, XX, XXX, zzz
Additionally S letter groups (fractional seconds) are supported
providing they occur after ss and separated from the ss by a dot,
comma or colon.  Spacing and punctuation is also permitted with
the exception of the question mark, newline and carriage return
characters, together with literal text enclosed in single quotes.

The full list of changes/improvements in this refactor is:

- Make TimestampFormatFinder an instantiable class
- Overrides must be specified in Java date/time format - Joda
  format is no longer accepted
- Joda timestamp formats in outputs are now derived from the
  determined or overridden Java timestamp formats, not stored
  separately
- Functionality for determining the "best" timestamp format in
  a set of lines has been moved from TextLogFileStructureFinder
  to TimestampFormatFinder, taking advantage of the fact that
  TimestampFormatFinder is now an instantiable class with state
- The functionality to quickly rule out some possible Grok
  patterns when looking for timestamp formats has been changed
  from using simple regular expressions to the much faster
  approach of using the Shift-And method of sub-string search,
  but using an "alphabet" consisting of just 1 (representing any
  digit) and 0 (representing non-digits)
- Timestamp format overrides are now much more flexible
- Timestamp format overrides that do not correspond to a built-in
  Grok pattern are mapped to a %{CUSTOM_TIMESTAMP} Grok pattern
  whose definition is included within the date processor in the
  ingest pipeline
- Grok patterns that correspond to multiple Java date/time
  patterns are now handled better - the Grok pattern is accepted
  as matching broadly, and the required set of Java date/time
  patterns is built up considering all observed samples
- As a result of the more flexible acceptance of Grok patterns,
  when looking for the "best" timestamp in a set of lines
  timestamps are considered different if they are preceded by
  a different sequence of punctuation characters (to prevent
  timestamps far into some lines being considered similar to
  timestamps near the beginning of other lines)
- Out-of-the-box Grok patterns that are considered now include
  %{DATE} and %{DATESTAMP}, which have indeterminate day/month
  ordering
- The order of day/month in formats with indeterminate day/month
  order is determined by considering all observed samples (plus
  the server locale if the observed samples still do not suggest
  an ordering)

Relates #38086
Closes #35137
Closes #35132

droberts195 committed:
Improve timestamp format quick-rule-out functionality

Previously if a timestamp format was not quickly ruled
out then we would search for it in the whole sample.
Following this change the quick-rule-out patterns are
used not only to completely rule out some formats but
also to find the portion of the sample over which the
format could possibly match.  This helps a lot in the
case of long lines that contain sections that nearly
match one of our candidate timestamps but not quite
(because regular expression matching is slowest in the
case of patterns that nearly match).

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +114,1074 @@
+    public boolean hasTimezoneDependentParsing() {
+        if (matchedFormats.isEmpty()) {
+            if (errorOnNoTimestamp) {
+                throw new IllegalStateException("No matched timestamp formats");
```
We should include the `why` in this exception.

droberts195 replied:
None of these `IllegalStateException`s in the getters are user facing error messages.  They mean the programmer has made a mistake by asking for results from an object they didn't add any samples to (because if they added a sample when `errorOnNoTimestamp` is `true` then _that_ would have thrown an exception).

It's obviously confusing as it is now.  I think I should change these exceptions to `assert`s instead since they're detecting usage errors by the programmer.  The return value for when assertions are disabled can just be the `null`/empty/false/0 that gets returned when `errorOnNoTimestamp` is `false`.  If you think that sounds OK then I'll change it tomorrow, but let me know if you think of a better approach.

benwtrent reviewed (commented):
Some minor things on the first read through. 

This is a ton to grok (pun intended, in fact, I think I used this same pun on the last time a huge PR was made for the fsf...).

droberts195 committed:
Address some code review comments

droberts195 committed:
Change exceptions for programmer mistakes to assertions

benwtrent reviewed (approved):
Will definitely need a second set of 👀 :)

droberts195 committed:
Finish off the custom Grok pattern functionality for custom timestamp formats

droberts195 committed:
Add missing Javadoc params

droberts195 committed:
Add support for ISO8601 date with no time

droberts195 committed:
Use empty map rather than null when no custom Grok patterns

droberts195 committed:
Bring secondary timestamps in Grok pattern creator in line with primary timestamps in timestamp format finder

droberts195 committed:
Merge branch 'master' into timestamp_finder_improvements

droberts195 committed:
Fix a couple of format quirks

1. Even though %{TIMESTAMP_ISO8601} cannot parse an ISO8601 date
   with no time, the ISO8601 date format can
2. The %{DATE} and %{DATESTAMP} Grok patterns accept a single
   digit month but not a single digit day

droberts195 committed:
Merge branch 'master' into timestamp_finder_improvements

droberts195 committed:
Fixing docs test

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/GrokPatternCreator.java:
```diff
@@ -486,9 +524,17 @@ public String processCaptures(Map<String, Integer> fieldNameCountStore, Collecti
-                    TimestampMatch timestampMatch = TimestampFormatFinder.findFirstFullMatch(values.iterator().next(), timeoutChecker);
-                    if (timestampMatch != null) {
-                        fullMappingType = timestampMatch.getEsDateMappingTypeWithFormat();
+                    TimestampFormatFinder timestampFormatFinder = new TimestampFormatFinder(explanation, true, true, true, timeoutChecker);
```
This for loop all the way to calling `timestampFormatFinder.getEsDateMappingTypeWithFormat()` is repeated in `FileStructureUtils` around line `277`. I wonder if we could refactor that in a method.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
-        return findFirstMatch(text, 0, timeoutChecker);
+    public TimestampFormatFinder(List<String> explanation, String overrideFormat, boolean requireFullMatch, boolean errorOnNoTimestamp,
+                                 boolean errorOnMultiplePatterns, TimeoutChecker timeoutChecker) {
+        this.explanation = explanation;
```
add `Objects.requireNonNull` or `@Nullable` where suitable

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
+        Tuple<String, String> grokPatternAndRegex = overrideFormatToGrokAndRegex(overrideFormat);
+        DateTimeFormatter javaTimeFormatter = DateTimeFormatter.ofPattern(overrideFormat, Locale.ROOT);
+
+        // This timestamp is chosen to avoid the possibility of truncating zeroes
```
I'm not quite getting why 29th Feb is doing the trick. Could you elaborate?

droberts195 replied:
Well spotted.  This is a bug.  The month, day and hour should all be less than 10 to distinguish formats that don't use leading zeroes for them.  I'll fix the comment too.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
-                for (Integer quickRuleOutIndex : candidate.quickRuleOutIndices) {
-                    if (quickRuleoutMatches[quickRuleOutIndex] == null) {
-                        quickRuleoutMatches[quickRuleOutIndex] = QUICK_RULE_OUT_PATTERNS.get(quickRuleOutIndex).matcher(text).find();
+    private static TimestampMatch checkCandidate(CandidateTimestampFormat candidate, String text, BitSet numberPosBitSet,
```
add `@Nullable` where suitable? Especially in methods with optional arguments it helps a lot when reading.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
+            return;
+        }
+
+        int remainingMatches = matches.size();
```
This block seems a nice candidate to extract into a method that returns the weights.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
+
+        timeoutChecker.check("timestamp format determination");
+
+        double highestWeight = 0.0;
```
This could be a `Tuple<Integer, Double> findHighestWeight(double[] weights)` method.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
+        timeoutChecker.check("timestamp format determination");
+
+        // Is the selected format not already at the beginning of the list?
+        if (highestWeightFormatIndex > 0) {
```
And finally, this could be `selectHighestWeightFormat`.

The above would make the method roughly read like:

```
calcMatchWeights();
findHighestWeight();
selectHighestWeightFormat();
```

Take all these as a mere suggestion. I know at the end it's down to personal preference, but I find breaking down methods like this increase readability a lot as it prepares the reader to understand what the lower-level code is trying to achieve.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -64,427 +113,1139 @@
+
+            if (onlyConsiderFormat == null || onlyConsiderFormat.canMergeWith(match.timestampFormat)) {
+
+                if (match.firstIndeterminateDateNumber > 0) {
```
What does the zero check represent here?

dimitris-athanasiou reviewed (commented):
Left a few comments. There are some cool but complex ideas in here!

droberts195 committed:
Memory optimisations

Also fixing a couple of typos

droberts195 committed:
Merge branch 'master' into timestamp_finder_improvements

droberts195 committed:
Address review comments

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java:
```diff
@@ -718,7 +793,11 @@ Boolean guessIsDayFirstFromMatches(TimestampFormat onlyConsiderFormat) {
 
             if (onlyConsiderFormat == null || onlyConsiderFormat.canMergeWith(match.timestampFormat)) {
 
+                // Valid indetermine day/month numbers will be in the range 1 to 31.
```
typo: `indeterminate`; it is a confusing one to type :-)

droberts195 replied:
Gah, I fixed the exact same typo in another place in another recent commit

dimitris-athanasiou reviewed (approved):
LGTM Just left a typo in a comment

droberts195 committed:
Fix typo

droberts195 commented:
Jenkins run elasticsearch-ci/1
